magento/magento2
# Cache Elasticsearch field mappers per entity

### Description (*)


Currently the Elasticsearch FieldMapperResolver caches the first resolved field mapper into a private class property called `$fieldMapperEntity`. All subsequents calls to the `getEntity` method in the resolver will return the field mapper which was initialized first, regardless of the supplied `$entityType` argument. This prevents different entity types (say, manufacturer names or categories) from using their own field mapper. 

### Manual testing scenarios (*)

1. Create a custom search indexer for Elasticsearch in Magento 2:
```
<indexer id="analysis_certificate_search_fulltext" view_id="analysis_certificate_search_fulltext" class="MyCompany\AnalysisSearch\Model\Indexer\FulltextAnalysisIndexer">
    <title translate="true">Analysis Certificate Search</title>
    <description translate="true">Rebuild analysis certificate fulltext search index</description>
</indexer>
```
2. Specify relevant configuration, but especially the custom field mapper for the indexer through `di.xml`:
```
<type name="Magento\Elasticsearch\Model\Adapter\FieldMapper\FieldMapperResolver">
    <arguments>
        <argument name="fieldMappers" xsi:type="array">
            <item name="analysis_certificate_search_fulltext" xsi:type="string">MyCompany\AnalysisSearch\Model\Elasticsearch\Adapter\AnalysisFieldMapper</item>
        </argument>
    </arguments>
</type>
```
3. Run the indexer in the same command with the product search index first:
```
$ bin/magento indexer:reindex catalogsearch_fulltext analysis_certificate_search_fulltext
```
4. Notice (through Kibana or otherwise) that the ES analysis certificate index is erroneously using the product index mapping. Get the mapping with:
```
GET /magento2_analysis_certificate_search_fulltext_1/_mapping
```

### Questions or comments


### Contribution checklist (*)
